pysradb/search.py

```diff
@@ -0,0 +1,437 @@
+"""This file contains the search classes for the search feature.
+"""
+
+import re
+import urllib
+import xml.etree.ElementTree as Et
+from collections import OrderedDict
+from json import JSONDecodeError
+
+import pandas as pd
+import requests
+
+from .exceptions import IncorrectFieldException
+from .exceptions import MissingQueryException
+from .utils import scientific_name_to_taxid, requests_3_retries
+
+
+class QuerySearch:
+    """This is the base class for the search feature.
+
+    This class takes as input the user's search query, which has been
+    tokenized by the ArgParser. The query will be sent to either SRA or ENA
+    depending on the user's input, and the results will be returned as a
+    pandas dataframe.
+
+    Attributes
+    ----------
+    self.df: Pandas DataFrame
+        The search result belonging to this search instance
+
+    Parameters
+    ----------
+    verbosity : integer
+        The level of details of the search result.
+    return_max : int
+        The maximum number of entries to be returned.
+    fields : dict
+        The fields supplied for the query. If all fields are empty, raises
+        MissingQueryException.
+
+    Methods
+    -------
+    get_df()
+        Returns the dataframe storing this search result
+
+    """
+
+    def __init__(self, verbosity, return_max, fields):
+        self.verbosity = verbosity
+        self.return_max = return_max
+        allowed_fields = ["query", "accession", "organism", "layout", "mbases", "publication_date", "platform",
+                          "selection", "source", "strategy", "title"]
+        for field in allowed_fields:
+            if field not in fields:
+                fields[field] = None
+        for k in fields:
+            if type(fields[k]) == list:
+                fields[k] = " ".join(fields[k])
+        if not any(fields):
+            raise MissingQueryException()
+        self.fields = fields
+        self.df = pd.DataFrame()
+
+    def search(self):
+        pass
+
+    def get_df(self):
+        return self.df
+
+
+class SraSearch(QuerySearch):
+    """Subclass of QuerySearch that implements search by querying
+    NCBI Entrez API
+
+    Methods
+    -------
+    search()
+        sends the user query via requests to NCBI Entrez API and returns
+        search results as a pandas dataframe.
+
+    _format_query_string()
+        formats the input user query into a string
+
+    _format_request()
+        formats the request payload
+
+    _format_result(content)
+        formats the search query output.
+
+    See Also
+    --------
+    QuerySearch: Superclass of SraSearch
+
+    """
+
+    def search(self):
+        # Step 1: retrieves the list of uids that satisfies the input
+        # search query
+
+        payload = self._format_request()
+        try:
+            r = requests_3_retries().get(
+                "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi",
+                params=OrderedDict(payload),
```

Comment:
I don't think `OrderedDict` adds anything here (dictionaries have stable order since python 3.6), and since `_format_request` just uses dictionary literals all sorting would be lost there anyway. Also it doesn't look like parameters need to be sorted in the entrez API.